src/lfortran/semantics/asr_implicit_cast_rules.h

```diff
@@ -184,7 +223,40 @@ class ImplicitCastRules {
                     value = (ASR::expr_t *)ASR::make_RealConstant_t(al, a_loc,
                         rval, dest_type2);
               } else {
-                    value = nullptr;
+                LCOMPILERS_ASSERT(ASR::is_a<ASR::ArrayConstant_t>(*value));
+                ASR::ArrayConstant_t* array = ASR::down_cast<ASR::ArrayConstant_t>(value);
+                ASR::Array_t* array_type = ASR::down_cast<ASR::Array_t>(array->m_type);
+                int kind = ASRUtils::extract_kind_from_ttype_t(array->m_type);
+                void *data = array->m_data;
+                size_t array_size = ASRUtils::get_fixed_size_of_array(array->m_type);
+                int dest_kind = ASRUtils::extract_kind_from_ttype_t(dest_type2);
+                if (dest_kind == 4) {
+                  float *new_array = new float[array_size];
+                  for (size_t i = 0; i < array_size; i++) {
+                    if ( kind == 4 ) {
+                      new_array[i] = float(((float*) data)[i]);
```

Comment:
I think you can just write this as:
```suggestion
                      new_array[i] = ((float*) data)[i];
```

Reply:
if `dest_kind == 4`, can `kind ==4 `?

Reply:
No, it can't, I will change it.